conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureError.java

```diff
@@ -0,0 +1,117 @@
+import com.palantir.conjure.java.api.errors.RemoteException;
+import com.palantir.conjure.java.api.errors.SerializableError;
+import com.palantir.conjure.java.api.errors.ServiceException;
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.Preconditions;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalDouble;
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+
+/**
+ * A representation of a Conjure defined error that will be serialized and sent over the wire from a server to a client.
+ * <p>Previously {@link SerializableError} was used for this purpose. The main difference is that
+ * {@link SerializableError#parameters()} is a {@code Map<String, String>} while {@link ConjureError#parameters} is a
+ * {@code Map<String, Object>}. When constructing a {@link SerializableError} from a {@link ServiceException}, the error
+ * parameters were converted to strings using {@link Objects#toString} which made it difficult for the client to recover
+ * the structured type. Storing the parameter value as an object allows servers to send Jackson-serialized objects which
+ * the client can convert to their original type.
+ *
+ * @param errorCode A fixed code word identifying the type of error. For errors generated from {@link ServiceException}
+ *                  and {@link CheckedServiceException}, this corresponds to the {@link ErrorType#code} and is part of
+ *                  the service's API surface.
+ * @param errorName A fixed name identifying the error. For errors generated from {@link ServiceException} and
+ *                  {@link CheckedServiceException}, this corresponding to the {@link ErrorType#name} and is part of the
+ *                  service's API surface.
+ * @param errorInstanceId errorInstanceId A unique identifier for (this instance of) this error.
+ * @param parameters params A list of parameters providing additional context about the error.
+ */
+record ConjureError(
+        @JsonProperty("errorCode") String errorCode,
+        @JsonProperty("errorName") String errorName,
+        @JsonProperty("errorInstanceId") String errorInstanceId,
+        @JsonInclude(content = JsonInclude.Include.CUSTOM, contentFilter = NonAbsentFilter.class)
+                Map<String, Object> parameters) {
+
+    static class NonAbsentFilter {
+        @Override
+        public boolean equals(Object obj) {
+            return obj == null
+                    || (obj instanceof Optional && ((Optional<?>) obj).isEmpty())
+                    || (obj instanceof OptionalInt && ((OptionalInt) obj).isEmpty())
+                    || (obj instanceof OptionalLong && ((OptionalLong) obj).isEmpty())
+                    || (obj instanceof OptionalDouble && ((OptionalDouble) obj).isEmpty());
+        }
+
+        @Override
+        public int hashCode() {
+            return getClass().hashCode();
+        }
+    }
```

Comment:
I see, given this, I think this will be easier to follow (and involve less reflection) if we do away with the `@JsonInclude` annotation in favor of filtering args that we place in the map (similar to your original implementation, but with additional instanceof checks for the other optional types).
Sorry for the runaround on this one!

Reply:
No worries! This was pretty non-obvious to me.